### DISCUSSION TITLE
Reason for the py::cast() error with custom BSDF?

### DISCUSSION THREAD

**Discussion by apoorvkhattar** [mitsuba-renderer/mitsuba3#317] [Q&A] [answered]
Hi again, 
I am trying to implement [irawan](https://github.com/mitsuba-renderer/mitsuba/blob/master/src/bsdfs/irawan.h) BSDF in python for mitsuba3 to familiarize myself with how to create custom plugins. I followed the C++ code carefully and the re-implemented plugin is working for a single test case with different SurfaceIntersection instances as input. However, when I try to render a sample XML file then I get the following error after 10-15s of rendering:

`When called outside a bound function, py::cast() cannot do Python -> C++ conversions which require the creation of temporary values`

I am not clear where this error happens, and whenever I have to cast a float to int, I am only casting a `drjit.float_array_t` to `dr.int_array_t` and there is no other time I explicitly require a cast operation.

If it can be clarified what is the reason for this error, I'd be grateful. I can share the code if more context is needed.

Entire error for reference:
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [9], in <cell line: 1>()
----> 1 image = mi.render(scene, seed=0)
      3 mi.Bitmap(image)

File ~\anaconda3\lib\site-packages\mitsuba\python\util.py:507, in render(scene, params, sensor, integrator, seed, seed_grad, spp, spp_grad)
    503 elif seed_grad == seed:
    504     raise Exception('The primal and differential seed should be different '
    505                     'to ensure unbiased gradient computation!')
--> 507 return dr.custom(_RenderOp, scene, sensor, params, integrator,
    508                  (seed, seed_grad), (spp, spp_grad))

File ~\anaconda3\lib\site-packages\drjit\router.py:5614, in custom(cls, *args, **kwargs)
   5612     output = inst.eval(*_dr.detach(kwargs['args']))
   5613 else:
-> 5614     output = inst.eval(**{ k: _dr.detach(v) for k, v in kwargs.items() })
   5615 if _dr.grad_enabled(output):
   5616     raise RuntimeError("drjit.custom(): the return value of CustomOp.eval() "
   5617                        "should not be attached to the AD graph!")

File ~\anaconda3\lib\site-packages\mitsuba\python\util.py:362, in _RenderOp.eval(self, scene, sensor, params, integrator, seed, spp)
    359 self.spp = spp
    361 with dr.suspend_grad():
--> 362     return self.integrator.render(
    363         scene=self.scene,
    364         sensor=sensor,
    365         seed=seed[0],
    366         spp=spp[0],
    367         develop=True,
    368         evaluate=False
    369     )

RuntimeError: When called outside a bound function, py::cast() cannot do Python -> C++ conversions which require the creation of temporary values


```

**Comment by Speierers** [comment, marked as answer]
Hi @apoorvkhattar ,

This means that C++ is not able to digest a Python object provided as an argument to a C++ function / method. If I understood correctly, you are implementing your own BSDF as a python plugin, similar to the [custom python plugin tutorial](https://mitsuba.readthedocs.io/en/stable/src/others/custom_plugin.html) is that right? 

Ideally if you could share the code for this plugin, we could help you debug this further. I suggest you check the inputs/outputs of the methods defined in this trampoline class, as they need to match the ones of the BSDF API.

**Reply by apoorvkhattar** [reply]
Hi 
I followed your advice to see if any wrong type was being returned in the sample() function of my BSDF and in one of the if conditions it was not returning value of type mi.Color hence the py::cast error was thrown. I changed that and it worked.

Thank you for your help.
